### PR TITLE
External Deletion for clusters

### PR DESCRIPTION
**Pull request by worryg0d**
Implemented handling of the external deletion of resources. In the previous implementation if there was no resource on the Instaclustr we deleted it in k8s. Now it changes a resource state to `DELETED` without deleting it in k8s.

### CODE REVIEW

File: controllers/clusters/helpers.go [hunk 1]
```diff
@@ -17,8 +17,11 @@ limitations under the License.
 package clusters
 
 import (
+	"context"
 	"encoding/json"
 	"fmt"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/log"
```
Comment (OleksiienkoMykyta) [inline controllers/clusters/helpers.go:24]: Please, sort imports, alse delete isClusterActive function on row 141

**Review by OleksiienkoMykyta — CHANGES_REQUESTED** [review]

File: controllers/clusters/helpers.go (hunk 1 above)
Comment (worryg0d) [reply controllers/clusters/helpers.go]: Fixed

**Review by OleksiienkoMykyta — APPROVED** [review]

**Review by tengu-alt — APPROVED** [review]

**Review by testisnullus — APPROVED** [review]

File: controllers/clusters/cassandra_controller_test.go [hunk 2]
```diff
@@ -106,4 +109,53 @@ var _ = Describe("Cassandra Controller", func() {
...
+				}
+
+				doneCh <- struct{}{}
+
+				return true
+			}, timeout, interval).Should(BeTrue())
+
+			<-doneCh
+
+			By("testing by deleting the cluster by Instaclustr API client")
```
Comment (ribaraka) [inline controllers/clusters/cassandra_controller_test.go:145]: By("using all possible ways other than k8s to delete a resource, the k8s operator scheduler should notice the changes and set the status of the deleted resource accordingly.")

File: controllers/clusters/kafkaconnect_controller_test.go [hunk 3]
```diff
@@ -18,6 +18,8 @@ package clusters
 
 import (
 	"context"
+	"github.com/instaclustr/operator/pkg/instaclustr"
+	"github.com/instaclustr/operator/pkg/models"
 	"os"
```
Comment (testisnullus) [inline controllers/clusters/kafkaconnect_controller_test.go:23]: Sort imports please

**Review by testisnullus — CHANGES_REQUESTED** [review]

File: controllers/clusters/cassandra_controller_test.go [hunk 4]
```diff
@@ -106,4 +109,53 @@ var _ = Describe("Cassandra Controller", func() {
...
+			Expect(instaClient.DeleteCluster(cassandra2.Status.ID, instaclustr.CassandraEndpoint)).Should(Succeed())
+
+			Eventually(func() bool {
+				err := k8sClient.Get(ctx, cassandra2NamespacedName, &cassandra2)
+				if err != nil {
+					return false
+				}
+
+				fmt.Println("cassandra2.Status.State", cassandra2.Status.State)
```
Comment (ribaraka) [inline controllers/clusters/cassandra_controller_test.go:154]: pls remove it

File: controllers/clusters/helpers.go [hunk 5]
```diff
@@ -22,11 +22,10 @@ import (
...
 
 	"github.com/hashicorp/go-version"
-	"k8s.io/utils/strings/slices"
-	"sigs.k8s.io/controller-runtime/pkg/client"
-
 	"github.com/instaclustr/operator/apis/clusters/v1beta1"
 	"github.com/instaclustr/operator/pkg/models"
+	"k8s.io/utils/strings/slices"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 )
```
Comment (ribaraka) [inline controllers/clusters/helpers.go:29]: imports

File: controllers/clusters/kafka_controller_test.go [hunk 6]
```diff
@@ -58,7 +59,7 @@ var _ = Describe("Kafka Controller", func() {
 					return false
 				}
 
-				return kafka.Status.ID == openapi.CreatedID
+				return kafka.Status.ID != ""
```
Comment (ribaraka) [inline controllers/clusters/kafka_controller_test.go:62]: why have you changed this?

File: controllers/clusters/kafkaconnect_controller.go [hunk 7]
```diff
@@ -19,7 +19,6 @@ package clusters
 import (
 	"context"
 	"errors"
-
 	"github.com/go-logr/logr"
```
Comment (ribaraka) [inline controllers/clusters/kafkaconnect_controller.go:22]: imports

File: controllers/clusters/kafkaconnect_controller_test.go [hunk 8]
```diff
@@ -58,10 +59,26 @@ var _ = Describe("Kafka Connect Controller", func() {
 					return false
 				}
 
-				return kafkaConnect.Status.ID == openapi.CreatedID
+				return kafkaConnect.Status.ID != ""
```
Comment (ribaraka) [inline controllers/clusters/kafkaconnect_controller_test.go:62]: this changes also seems unnecessary

File: controllers/clusters/kafkaconnect_controller_test.go [hunk 9]
```diff
@@ -58,10 +59,26 @@ var _ = Describe("Kafka Connect Controller", func() {
...
 			}).Should(BeTrue())
 
 			<-done
+
+			By("creating secret with the default user credentials")
+			secret := kafkaConnect.NewDefaultUserSecret("", "")
+			secretNamespacedName := types.NamespacedName{
+				Namespace: secret.Namespace,
+				Name:      secret.Name,
+			}
```
Comment (ribaraka) [inline controllers/clusters/kafkaconnect_controller_test.go:72]: is that OK to create a `defaultUserSecret` with empty creds?

File: pkg/instaclustr/mock/server/go/api_apache_kafka_connect_provisioning_v2_service.go [hunk 10]
```diff
@@ -79,7 +87,9 @@ func (s *ApacheKafkaConnectProvisioningV2APIService) ClusterManagementV2Resource
 	// TODO - update ClusterManagementV2ResourcesApplicationsKafkaConnectClustersV2Post with the required logic for this service method.
 	// Add api_apache_kafka_connect_provisioning_v2_service.go to the .openapi-generator-ignore to avoid overwriting this service implementation when updating open api generation.
 
-	s.MockKafkaConnectCluster = &kafkaConnectClusterV2
-	s.MockKafkaConnectCluster.Id = CreatedID
-	return Response(202, s.MockKafkaConnectCluster), nil
+	id := uuid.New().String()
+	kafkaConnectClusterV2.Id = id
+	s.clusters[id] = &kafkaConnectClusterV2
```
Comment (ribaraka) [inline pkg/instaclustr/mock/server/go/api_apache_kafka_connect_provisioning_v2_service.go:92]: I think it's better to use the ID setting logic that I implemented in the stub server API service for the Cassandra resource.

**Review by ribaraka — CHANGES_REQUESTED** [review]

File: pkg/instaclustr/mock/server/go/api_apache_kafka_connect_provisioning_v2_service.go (hunk 10 above)
Comment (ribaraka) [reply pkg/instaclustr/mock/server/go/api_apache_kafka_connect_provisioning_v2_service.go:92]: more control, imo

File: controllers/clusters/kafkaconnect_controller_test.go (hunk 9 above)
Comment (worryg0d) [reply controllers/clusters/kafkaconnect_controller_test.go:72]: I use it only for getting a secret with an already configured namespaced name for the specified cluster and nothing else.

File: controllers/clusters/cassandra_controller_test.go (hunk 4 above)
Comment (worryg0d) [reply controllers/clusters/cassandra_controller_test.go:154]: Removed.

File: controllers/clusters/cassandra_controller_test.go (hunk 2 above)
Comment (worryg0d) [reply controllers/clusters/cassandra_controller_test.go:145]: Fixed

File: controllers/clusters/kafkaconnect_controller_test.go (hunk 3 above)
Comment (worryg0d) [reply controllers/clusters/kafkaconnect_controller_test.go:23]: Fixed

File: controllers/clusters/helpers.go (hunk 5 above)
Comment (worryg0d) [reply controllers/clusters/helpers.go:29]: Fixed

File: controllers/clusters/kafkaconnect_controller.go (hunk 7 above)
Comment (worryg0d) [reply controllers/clusters/kafkaconnect_controller.go:22]: What's wrong here?

File: controllers/clusters/kafkaconnect_controller_test.go [hunk 11]
```diff
@@ -18,6 +18,8 @@ package clusters
 
 import (
 	"context"
+	"github.com/instaclustr/operator/pkg/instaclustr"
+	"github.com/instaclustr/operator/pkg/models"
 	"os"
 
 	. "github.com/onsi/ginkgo/v2"
```
[the remark this answers is not in the document]
Comment (worryg0d) [reply controllers/clusters/kafkaconnect_controller_test.go:25]: w_w

File: pkg/instaclustr/mock/server/go/api_apache_kafka_connect_provisioning_v2_service.go (hunk 10 above)
Comment (worryg0d) [reply pkg/instaclustr/mock/server/go/api_apache_kafka_connect_provisioning_v2_service.go:92]: It makes sense. I just forgot to revert these changes.

File: controllers/clusters/kafka_controller_test.go (hunk 6 above)
Comment (worryg0d) [reply controllers/clusters/kafka_controller_test.go:62]: Because I used uuid to generate IDs for entities. Reverted

**Review by ribaraka — APPROVED** [review]

**Review by testisnullus — APPROVED** [review]

**Review by taaraora — APPROVED** [review]

**Pull request merged**